// 2pitFlow/controller/FlowAccountController.js
const AccountTable = require('2pitCore/model/Account.js');
const PlaceTable = require('2pitCore/model/Place.js');
const VcardTable = require('2pitCore/model/Vcard.js');
const Sql = require('2pitCore/model/Sql');
const or = Sql.or;
const and = Sql.and;
const not = Sql.not;

exports.FlowAccount_fill = function (request, response, context, connector) {

  return new Promise( (resolve, reject) => {

    let type = request.params.type;
    let place, account, vcard;
    let locale = request.query.locale;
    
    // Retrieve the context place
    let connection = connector();
    PlaceTable.get(context, connection, context.instance.default_place_id)
    .then( pl => {
      place = pl;
      return AccountTable.get(context, connection, request.params.id)
    })
    .then( ac => {
      account = ac;
      return VcardTable.get(context, connection, account.contact_1_id);
    })
    .then ( vc => {
      connection.end( err => { if (err) throw err; } );
      vcard = vc;
      if (!locale) locale = vcard.locale;
      
      if (!account) account = AccountTable.instanciate(context, type);
      
      resolve({
        context: context,
        type: type, 
        locale: locale,
        place_identifier: place.identifier, 
        account: (account) ? account : '', 
      });
    })
    .catch( err => { reject(err); });
  });
};
